### PR TITLE
Build Haskell backend as a submodule, hook into Maven

### PR DESCRIPTION
**Pull request by ehildenb**
This is in preparation of running the K tests on all the backends, not just the Java backend.

It will help root out deficiencies in the Haskell backend.

### CODE REVIEW

**Comment by ehildenb** [comment]
Note that this should wait for: https://github.com/kframework/kore/pull/393

Once that is merged, I'll update the submodule and this will be ready.

**Review by traiansf — COMMENTED** [review]
Looks good, but I think some of the default values here: `org/kframework/backend/haskell/HaskellKRunOptions.java`  need to change as well.

**Comment by traiansf** [comment]
Also, it would be nice to add a test that the haskell backend works, or maybe even calling `make k-test` in the `kore` repo.

**Comment by ehildenb** [comment]
@traiansf I'm holding off the changes for making sure the Haskell backend works for another PR. My plan is to also make it possible to run the test-set in k-distribution with any of the backends, then to have a list of tests to run with each backend. Currently we just run some of them with OCaml and others with Java.

The Haskell backend integration tests will be tested on it's PR job for now, so there I think I'm going to focus on running as many of the tests in k-distribution that the Haskell backend can in a reasonable amount of time.

I also plan on making a weekly build job that run all integration tests on nightly builds to find a nightly release which is working for all the semantics we care about. Once that build is up, the Haskell backend PR will use that version for testing instead of the nightly.

**Comment by ehildenb** [comment]
Also, what defaults in HaskellKRunOptions need to change? We discussed on Slack, and I think the conclusion was that the current defaults actually work well. Right now it will default to using the binaries bundled in `k-distribution/target/release/k/bin`, and the binaries built by the haskell backend are being copied there.

You can override this by explicitly calling the options, which the testing for the Haskell backend needs. But the default is to use the version pinned by the submodule.

**Comment by ehildenb** [comment]
Jenkins: test this please

File: haskell-backend/pom.xml
Comment (dwightguth) [inline haskell-backend/pom.xml]: I'm reasonably sure this is not necessary as I added this flag specifically to deal with cmake.

**Review by dwightguth — CHANGES_REQUESTED** [review]
looks good except one minor issue but I agree this should be tested.

File: haskell-backend/pom.xml
Comment (ehildenb) [reply haskell-backend/pom.xml]: Removed.

**Comment by ehildenb** [comment]
Can one of the admins verify this patch?

**Comment by ehildenb** [comment]
@dwightguth this is ready for re-review.

I've parameterized the k-distribution integration tests in the backend to run them in (Makefile variable `KOMPILE_BACKEND`). I've also switched `1_k/2_imp` over to using the Haskell backend by default, to prevent regressions.

File: k-distribution/tutorial/1_k/2_imp/Makefile
Comment (dwightguth) [inline k-distribution/tutorial/1_k/2_imp/Makefile]: This is a bad choice as it's going to mess with the students building the tutorial in the next couple weeks and expose them to any lingering bugs in the Haskell backend. You should put the regression test for the Haskell backend in tests/regression-new

**Review by dwightguth — CHANGES_REQUESTED** [review]
Minor change needed but the code still looks good

File: k-distribution/tutorial/1_k/2_imp/Makefile
Comment (ehildenb) [reply k-distribution/tutorial/1_k/2_imp/Makefile]: Which regression tests are you referring to?

Comment (dwightguth) [reply k-distribution/tutorial/1_k/2_imp/Makefile]: that's the name of the folder: k-distribution/tests/regression-new

Comment (ehildenb) [reply k-distribution/tutorial/1_k/2_imp/Makefile]: Alright, I've made a new regression testsuite imp-haskell which exercises the haskell backend, leaving the tutorial alone.

**Review by dwightguth — APPROVED** [review]

**Comment by dwightguth** [comment]
I approved but I'd like Traian to approve too before this gets merged.

**Comment by ehildenb** [comment]
@dwightguth I've added an option `-Dhaskell.backend.skip`, tested it locally and it seems to work.

File: k-distribution/include/ktest-keq.mak
Comment (traiansf) [inline k-distribution/include/ktest-keq.mak:5]: Don't you need a ? here to allow changing the backend?

**Review by traiansf — APPROVED** [review]

**Comment by traiansf** [comment]
Nevermind my comment; I saw some tests have fixed backend.

File: k-distribution/include/ktest-keq.mak
Comment (ehildenb) [reply k-distribution/include/ktest-keq.mak:5]: The keq testset always uses the Java backend for now. It's a different execution mode.

**Pull request merged**
